inc/gantt/datafactory.class.php

```diff
@@ -0,0 +1,164 @@
+ * (at your option) any later version.
+ *
+ * GLPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\Gantt;
+
+use Glpi\Gantt\Item;
+use Glpi\Gantt\LinkDAO;
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+/**
+ * Class used to prepare data for Gantt
+ */
+class DataFactory {
+
+   /**
+    * Recursive function used to get all subitems of a project, when $id > 0.
+    * Returns all projects with their subitems if $id == -1 (for global gantt view).
+    *
+    * @param array $itemArray Array holding the result
+    * @param integer $id ID of the parent project
+    */
+   function getItemsForProject(&$itemArray, $id) {
+      global $DB;
+      $project = new \Project();
+      if ($id == - 1) {
+         $iterator = $DB->request([
+            'FROM' => 'glpi_projects',
+            'WHERE' => [
+               'projects_id' => 0,
+               'show_on_global_gantt' => 1,
+               'is_template' => 0,
+               'is_deleted' => 0
+            ] + getEntitiesRestrictCriteria('glpi_projects', '', '', true)
+         ]);
+         while ($data = $iterator->next()) {
+            $this->getItemsForProject($itemArray, $data['id']);
+         }
+      }
+      else if ($project->getFromDB($id)) {
+         array_push($itemArray, $this->populateGanttItem($project->fields, "root-project"));
+         $this->getProjectTasks($itemArray, $id);
+         $this->getSubprojects($itemArray, $id); // subproject tasks included
+      }
+   }
+
+   /**
+    * Function used to get project task links
+    *
+    * @param array $itemArray Input array holding project and task items
+    *
+    * @return array $links Array of Link objects
+    */
+   function getProjectTaskLinks($itemArray) {
+      $links = [];
+      if (isset($itemArray)) {
+
+         $ids = [];
+         foreach ($itemArray as $item) {
+            if ($item->type != 'project')
+               $ids[] = $item->linktask_id;
+         }
+
+         if (count($ids) > 0) {
+            $linkDao = new LinkDAO();
+            $links = $linkDao->getLinksForItemIDs($ids);
+         }
+      }
+      return $links;
+   }
+
+   /**
+    * Recursive function used to get all subprojects and tasks of a project
+    *
+    * @param array $itemArray Array holding the items
+    * @param integer $projectId ID of the parent project
+    *
+    */
+   function getSubprojects(&$itemArray, $projectId) {
+      global $DB;
+      foreach ($DB->request('glpi_projects', ['projects_id' => $projectId, 'is_deleted' => 0]) as $record) {
```

Comment:
For performance reasons, avoid calling "complex" methods in any loop
```suggestion
      $iterator = $DB->request('glpi_projects', ['projects_id' => $projectId, 'is_deleted' => 0]);
      while ($record = $iterator->next()) {
```